src/runtime-rs/crates/hypervisor/src/device/device_manager.rs

```diff
@@ -0,0 +1,252 @@
+        self.released_block_index.push(index);
+        self.released_block_index.sort_by(|a, b| b.cmp(a));
+    }
+}
+
+// Device manager will manage the lifecycle of sandbox device
+pub struct DeviceManager {
+    devices: HashMap<String, ArcMutexBoxDevice>,
+    hypervisor: Arc<dyn Hypervisor>,
+    shared_info: SharedInfo,
+}
+
+impl DeviceManager {
+    pub fn new(hypervisor: Arc<dyn Hypervisor>) -> Result<Self> {
+        let devices = HashMap::<String, ArcMutexBoxDevice>::new();
+        Ok(DeviceManager {
+            devices,
+            hypervisor,
+            shared_info: SharedInfo::new(),
+        })
+    }
+
+    pub async fn new_device(&mut self, device_config: &DeviceConfig) -> Result<String> {
+        let device_id = if let Some(dev) = self.find_device(device_config).await {
+            dev
+        } else {
+            self.create_device(device_config)
+                .await
+                .context("failed to create device")?
+        };
+        Ok(device_id)
+    }
+
+    pub async fn try_add_device(&mut self, device_id: &str) -> Result<()> {
+        let device = self
+            .devices
+            .get(device_id)
+            .context("failed to find device")?;
+        let mut device_guard = device.lock().await;
+        // increase attach count, skip attach the device if the device is already attached
+        let need_skip = device_guard
+            .increase_attach_count()
+            .await
+            .context("failed to increase attach count")?;
+        if need_skip {
+            return Ok(());
+        }
+        let result = device_guard.attach(self.hypervisor.as_ref()).await;
+        // handle attach error
+        if let Err(e) = result {
+            device_guard.decrease_attach_count().await?;
+            if let DeviceConfig::Block(config) = device_guard.get_device_info().await {
+                self.shared_info.release_device_index(config.index);
+            };
+            drop(device_guard);
+            self.devices.remove(device_id);
+            return Err(e);
+        }
+        Ok(())
+    }
+
+    pub async fn try_remove_device(&mut self, device_id: &str) -> Result<()> {
+        if let Some(dev) = self.devices.get(device_id) {
+            let mut device_guard = dev.lock().await;
+            // get the count of device detached, skip detach once it reaches the 0.
+            let skip = device_guard.decrease_attach_count().await?;
+            if skip {
+                return Ok(());
+            }
+            let result = match device_guard.detach(self.hypervisor.as_ref()).await {
+                Ok(index) => {
+                    if let Some(i) = index {
+                        // release the declared block device index
+                        self.shared_info.release_device_index(i);
+                    }
+                    Ok(())
+                }
+                Err(e) => {
+                    device_guard.increase_attach_count().await?;
```

Comment:
This handling method is not so good for me. 
Currently, it seems like in `device.attach()`, you first decrease the attach_count and then do the truly detach device. IMO, I think you should decrease attach_count after ensuring detaching successfully in `device.detach()`.

Reply:
We can not do detach first, as the operation of detach device is decided by attach count, if the attach count is greater than 1, we cannot detach.

Reply:
Thanks, I have put the decrease_attach_count/increase_attach_count operation inside the attach/detach method to keep atomic operation.